docs/interfaces.md

```diff
@@ -0,0 +1,571 @@
+Depending on the attestation protocol(e.g., EPID-based attestation), the verifier implementes the corresponding logic.
+
+```go
+type EnclaveRegistryCC struct {
+	ra  attestation.Verifier
+}
+
+```
+
+
+# TLCC
+
+Some notes on TLCC. The TLCC instance is currently implemented as a systemchaincode, that is, there exists only a single instance. As a peer can participate in many channels, there is a seperate TLCC_enclave per each channel. The TLCC chaincode is responsible to multiplex requests for a given channel to the corresponding TLCC_enclave. Note for MVP, we only support a single channel but the interface specified here should already take multi-channel support into account.
+
+## Interface:
+The TLCC interface is implemented through a "normal" chaincode interface.
+
+```go
+// chaincode interface
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// functions below are dispatched by Invoke implementation
+
+// used by the admin to join the channel
+// note that the implementation of this method needs to retrieve the
+// genesis and all subsequence blocks for the given channel and pass
+// it to TLCC_enclave.
+func joinChannel(channel_id string) error {}
+
+// interface to establish/close a secure channel between an
+// ECC_Enclave and a TLCC_enclave.
+func session_setup(setup SessionSetup) (SessionSetupResponse, error) {}
+func session_close(close SessionClose) (SessionCloseReponse, error) {}
+
+// interface to invoke a TLCC request, such as, getMetadata and validateMSPId
+func session_request(request SessionRequest) (SessionRequestResponse, error) {}
+```
+
+```c++
+// Provided/implemented by a common logging module
+// interface exposed to TLCC_enclave via cgo
+void ocall_print_string([in, string] const char *str);
+```
+
+## State:
+TLCC keeps a local reference of the enclave id. This reference is volatile. Also a reference to the peer is kept to access the ledger and read blocks.
+
+```go
+type TrustedLedgerCC struct {
+	enclave enclave.Stub
+	peer    *peer.Peer
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+```protobuf
+syntax = "proto3"
+
+// Setup and Close messages are used to establish a secure channel between
+// ECC_Enclave and TLCC_Enclave. Once a secure channel is established,
+// Session Requests to TLCC can be issued to invoke tlcc functions such
+// as getMetadata or verifyMSPId.
+
+// TODO How to integrate the the message authenticator?!
+// should we abstract it as signature here even though we may
+// use cmac once a session is established?
+
+message SessionSetup {
+        // TBD input enclave vk?
+        // TLCC Mrenclave?
+        // channel
+}
+
+message SessionSetupResponse {
+        SessionSetup setup = 1;
+        byte context       = 2;         // context that references the session
+}
+
+message SessionClose {
+        byte context = 1;
+}
+
+message SessionCloseResponse {
+}
+
+message SessionRequest {
```

Comment:
If we do on session for whole lifetime of ecc, we also need here a transaction identifier so we can make sure all state returned for a single transactions are all based on the same ledger state